File: web/src/lib/resource-counts.ts
import type { BootstrapData, IssueLabel } from '@/types/flow'

export function deriveResourceCounts(data: BootstrapData): BootstrapData {
  if (data.issueCollectionPaged) return { ...data, issues: data.issues.slice(0, 2000) }
  const labelCounts = new Map<string, number>()
  const projectIssueCounts = new Map<string, number>()
  const increment = (ids: string[]) => {
    for (const id of new Set(ids.filter(Boolean))) labelCounts.set(id, (labelCounts.get(id) ?? 0) + 1)
  }

  for (const issue of data.issues) {
    if (issue.archivedAt) continue
    increment(issue.labels.map(label => label.id))
    if (issue.project) projectIssueCounts.set(issue.project.id, (projectIssueCounts.get(issue.project.id) ?? 0) + 1)
  }
  for (const project of data.projects) {
    if (!project.archivedAt) increment(project.labelIds ?? [])
  }
  for (const initiative of data.initiatives) increment(initiative.labelIds ?? [])

  const labels = data.labels.map(label => { const count = labelCounts.get(label.id) ?? 0; return label.issueCount === count ? label : { ...label, issueCount: count } })
  const labelsById = new Map<string, IssueLabel>(labels.map(label => [label.id, label]))
  return {
    ...data,
    labels,
    issues: data.issues.map(issue => {
      if (issue.labels.every(label => !labelsById.has(label.id) || labelsById.get(label.id) === label)) return issue
      return { ...issue, labels: issue.labels.map(label => labelsById.get(label.id) ?? label) }
    }),
    projects: data.projects.map(project => { const count = projectIssueCounts.get(project.id) ?? 0; return project.issueCount === count ? project : { ...project, issueCount: count } }),
  }
}

File: api/internal/store/resource_counts.go
package store

import "flow/api/internal/domain"

// refreshResourceCounts derives aggregate counts from the relationships that
// own them. Stored counters are only denormalized read models.
func refreshResourceCounts(data *domain.Bootstrap) {
	counts := make(map[string]int, len(data.Labels))
	projectIssueCounts := make(map[string]int, len(data.Projects))

	for _, issue := range data.Issues {
		if issue.ArchivedAt != nil {
			continue
		}
		incrementLabelCounts(counts, labelIDs(issue.Labels))
		if issue.Project != nil {
			projectIssueCounts[issue.Project.ID]++
		}
	}
	for _, project := range data.Projects {
		if project.ArchivedAt != nil {
			continue
		}
		incrementLabelCounts(counts, project.LabelIDs)
	}
	for _, initiative := range data.Initiatives {
		incrementLabelCounts(counts, initiative.LabelIDs)
	}

	labelsByID := make(map[string]domain.IssueLabel, len(data.Labels))
	for index := range data.Labels {
		data.Labels[index].IssueCount = counts[data.Labels[index].ID]
		labelsByID[data.Labels[index].ID] = data.Labels[index]
	}
	for issueIndex := range data.Issues {
		for labelIndex := range data.Issues[issueIndex].Labels {
			if label, ok := labelsByID[data.Issues[issueIndex].Labels[labelIndex].ID]; ok {
				data.Issues[issueIndex].Labels[labelIndex] = label
			}
		}
	}
	for index := range data.Projects {
		data.Projects[index].IssueCount = projectIssueCounts[data.Projects[index].ID]
	}
}

func incrementLabelCounts(counts map[string]int, ids []string) {
	seen := make(map[string]struct{}, len(ids))
	for _, id := range ids {
		if id == "" {
			continue
		}
		if _, ok := seen[id]; ok {
			continue
		}
		seen[id] = struct{}{}
		counts[id]++
	}
}

func labelIDs(labels []domain.IssueLabel) []string {
	ids := make([]string, 0, len(labels))
	for _, label := range labels {
		ids = append(ids, label.ID)
	}
	return ids
}
